runtime/auth/aws-signing-standard/common/src/aws/smithy/kotlin/runtime/auth/awssigning/standard/SignatureCalculator.kt

```diff
@@ -0,0 +1,93 @@
+     * @param prevSignature The signature of the previous chunk. If this is the first chunk, use the seed signature.
+     * @param config The signing configuration to use
+     * @return A multiline string to sign
+     */
+    fun chunkStringToSign(chunkBody: ByteArray, prevSignature: ByteArray, config: AwsSigningConfig): String
+
+    /**
+     * Derives a signing key
+     * @param config The signing configuration to use
+     * @param credentials Retrieved credentials to use
+     * @return The signing key as a byte array
+     */
+    fun signingKey(config: AwsSigningConfig, credentials: Credentials): ByteArray
+
+    /**
+     * Constructs a string to sign for a request
+     * @param canonicalRequest The canonical request string (returned from [Canonicalizer.canonicalRequest])
+     * @param config The signing configuration to use
+     * @return A multiline string to sign
+     */
+    fun stringToSign(canonicalRequest: String, config: AwsSigningConfig): String
+}
+
+internal class DefaultSignatureCalculator(private val sha256Provider: HashSupplier = ::Sha256) : SignatureCalculator {
+    private val logger = Logger.getLogger<DefaultSignatureCalculator>()
+
+    override fun calculate(signingKey: ByteArray, stringToSign: String): String =
+        hmac(signingKey, stringToSign.encodeToByteArray(), sha256Provider).encodeToHex()
+
+    override fun chunkStringToSign(chunkBody: ByteArray, prevSignature: ByteArray, config: AwsSigningConfig): String =
+        buildString {
+            appendLine("AWS4-HMAC-SHA256-PAYLOAD")
+            appendLine(config.signingDate.format(TimestampFormat.ISO_8601_CONDENSED))
+            appendLine(config.credentialScope)
+            appendLine(prevSignature.decodeToString()) // Should already be a byte array of ASCII hex chars
+            appendLine(BodyHash.EmptyBody.hash)
+            append(chunkBody.hash(sha256Provider).encodeToHex())
+        }.also { logger.debug { "String to sign:\n$it" } }
+
+    override fun signingKey(config: AwsSigningConfig, credentials: Credentials): ByteArray {
+        fun hmac(key: ByteArray, message: String) = hmac(key, message.encodeToByteArray(), sha256Provider)
+
+        val initialKey = ("AWS4" + credentials.secretAccessKey).encodeToByteArray()
+        val kDate = hmac(initialKey, config.signingDate.format(TimestampFormat.ISO_8601_CONDENSED_DATE))
+        val kRegion = hmac(kDate, config.region)
+        val kService = hmac(kRegion, config.service)
+        return hmac(kService, "aws4_request")
+    }
+
+    override fun stringToSign(canonicalRequest: String, config: AwsSigningConfig): String =
+        buildString {
+            appendLine("AWS4-HMAC-SHA256")
+            appendLine(config.signingDate.format(TimestampFormat.ISO_8601_CONDENSED))
+            appendLine(config.credentialScope)
+            append(canonicalRequest.encodeToByteArray().hash(sha256Provider).encodeToHex())
+        }.also { logger.debug { "String to sign:\n$it" } }
```

Comment:
question: Should signing related logs be gated by an explicit `SdkLogMode` ? 🤔

Reply:
They're not gated in CRT nor in Java v2 by anything other than logging level. I don't believe the things we're logging (or would ever log) include sensitive information...just intermediate state. So my instinct was no additional gating is necessary but happy to counterarguments.

Reply:
Nope more of a sanity check that we gave it consideration